pkg/metrics/sampler/matcher.go

```diff
@@ -277,70 +262,78 @@ func (ne constantMatcher) String() string {
-	// configuration option always takes precedence over FF and matchers configuration
-	if enableProcessMetrics == nil {
-		matcher := matcherFromMetricsMatchers(metricsMatchers)
-		if matcher != nil {
-			return matcher
+// NewIncludeSampleMatchFn returns a function `func(sample) bool` that determinesif the sample
+// should be included (true) as an event or not (false). Note that this is NOT the negation
+// of `NewExcludeSampleMatchFn`.
+// The include decision logic only applies to ProcessSamples. Other kinds of samples are always included.
+func NewIncludeSampleMatchFn(enableProcessMetrics *bool, metricsMatchers config.IncludeMetricsMap, ffRetriever feature_flags.Retriever) MatcherFn {
+	return func(sample any) bool {
+		// We return early if the sample is not a ProcessSample.
+		if !isProcessSample(sample) {
+			// Include this event
+			return true
 		}
 
-		// configuration option is not defined and feature flag is present, FF determines, otherwise
-		// all process samples will be excluded
-		return matcherFromFeatureFlag(ffRetriever)
-	}
+		// Continuing with the logic
+		// configuration option always takes precedence over FF and matchers configuration
+		if enableProcessMetrics == nil {
+			matcher := matcherFromMetricsMatchers(config.MetricsMap(metricsMatchers))
+			if matcher != nil {
+				return matcher(sample)
+			}
 
-	if excludeProcessMetrics(enableProcessMetrics) {
-		mlog.
-			WithField(config.TracesFieldName, config.FeatureTrace).
-			Trace("EnableProcessMetrics is FALSE, process metrics will be DISABLED")
+			// configuration option is not defined and feature flag is present, FF determines, otherwise
+			// all process samples will be excluded. Note that this is not exactly the negation
+			// of `NewExcludeSampleMatchFn`.
+			enabled, exists := ffRetriever.GetFeatureFlag(fflag.FlagFullProcess)
 
-		return matcherForDisabledMetrics()
-	}
+			return exists && enabled
+		}
 
-	matcherChain := NewMatcherChain(metricsMatchers)
-	if matcherChain.Enabled {
-		mlog.
-			WithField(config.TracesFieldName, config.FeatureTrace).
-			Trace("EnableProcessMetrics is TRUE and rules ARE defined, process metrics will be ENABLED for matching processes")
+		if !*enableProcessMetrics {
+			mlog.
+				WithField(config.TracesFieldName, config.FeatureTrace).
+				Trace("EnableProcessMetrics is FALSE, process metrics will be DISABLED")
 
-		return matcherChain.Evaluate
-	}
+			return false
+		}
 
-	mlog.
-		WithField(config.TracesFieldName, config.FeatureTrace).
-		Trace("EnableProcessMetrics is TRUE and rules are NOT defined, ALL process metrics will be ENABLED")
+		matcher := matcherFromMetricsMatchers(config.MetricsMap(metricsMatchers))
+		if matcher != nil {
+			return matcher(sample)
+		}
+
+		mlog.
+			WithField(config.TracesFieldName, config.FeatureTrace).
+			Trace("EnableProcessMetrics is TRUE and rules are NOT defined, ALL process metrics will be ENABLED")
```

Comment:
quite a minor thing but this log will be printed en case also when exclude matchers are added and no include ones.

Reply:
Does the latest commit (https://github.com/newrelic/infrastructure-agent/pull/1943/commits/abdd6961729382184feb1424670eeb32af6fa701) solve this? I reworded a bit.